src/lib/logger/impl.ml

```diff
@@ -207,74 +217,79 @@ module Transport = struct
         t.primary_log_size <- t.primary_log_size + len
     end
 
-    let dumb_logrotate ~directory ~max_size =
-      T ((module Dumb_logrotate), Dumb_logrotate.create ~directory ~max_size)
+    let dumb_logrotate ~directory ~log_name ~max_size =
+      T
+        ( (module Dumb_logrotate)
+        , Dumb_logrotate.create ~directory ~log_name ~max_size )
   end
 end
 
 module Consumer_registry = struct
-  type id = string
-
   type consumer = {processor: Processor.t; transport: Transport.t}
 
-  type nonrec t = (id, consumer) List.Assoc.t ref
+  module Consumer_tbl = Hashtbl.Make (String)
 
-  let t : t = ref []
+  type t = consumer list Consumer_tbl.t
 
-  let is_registered id = Option.is_some (List.Assoc.find !t id ~equal:( = ))
+  let t : t = Consumer_tbl.create ()
 
   let register ~id ~processor ~transport =
-    if is_registered id then
-      failwith "cannot register logger consumer with the same id twice"
-    else t := List.Assoc.add !t id {processor; transport} ~equal:( = )
+    Consumer_tbl.add_multi t ~key:id ~data:{processor; transport}
 
-  let broadcast_log_message msg =
+  let broadcast_log_message ~id msg =
     (* TODO: warn or fail if there's no registered consumer? Issue #3000 *)
-    List.iter !t ~f:(fun (_id, consumer) ->
-        let (Processor.T ((module Processor_mod), processor)) =
-          consumer.processor
+    Hashtbl.find_and_call t id
+      ~if_found:(fun consumers ->
+        List.iter consumers
+          ~f:(fun { processor= Processor.T ((module Processor), processor)
+                  ; transport= Transport.T ((module Transport), transport) }
+             ->
+            match Processor.process processor msg with
+            | Some str ->
+                Transport.transport transport str
+            | None ->
+                () ) )
+      ~if_not_found:(fun _ ->
```

Comment:
does this make issue #3000 obsolete? if so, the comment above can be deleted, and the issue closed